File: batch/src/main/scala/jobs_batch.scala
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf
import org.apache.spark.sql.SQLContext
import org.apache.spark.sql._
import org.elasticsearch.spark._
import org.elasticsearch.spark.rdd.EsSpark

object jobs_batch {
   def main(args: Array[String]) {

      // setup the Spark Context named sc
      val conf = new SparkConf().setAppName("jobs_batch")
      conf.set("es.index.auto.create", "true")
      conf.set("es.resource", "dashboard/jobs")
      conf.set("es.nodes", "localhost:9200")
      //conf.set("es.nodes", "awsHost:9200")
      //conf.set("spark.serializer", classOf[KryoSerializer].getName)
      //conf.set("es.query", "?q=firstname:Daniel")

      //val sc = new SparkContext(conf)
      val sc = new SparkContext(conf)
      //.hadoopConfiguration.set("fs.s3a.access.key", sys.env.getOrElse("AWS_ACCESS_KEY_ID", sys.error("missing AWS_ACCESS_KEY_ID")
      //.hadoopConfiguration.set("fs.s3a.secret.key", sys.env.getOrElse("AWS_SECRET_ACCESS_KEY", sys.error("missing AWS_SECRET_ACCESS_KEY")

      val csqlContext = new SQLContext(sc)
      import csqlContext.implicits._
      
      // folder on HDFS to pull the data from
      val diceFile = "hdfs://ec2-52-89-46-245.us-west-2.compute.amazonaws.com:9000/camus/exec/history/2016-09-16*"
      val diceFileS3 = "s3a://giovanna-insight/raw_logs/secor_backup/diceFeed/"
      

      // construct RDD[Sting]
      val indeedStaticFile = sc.parallelize(
         """{"snippet": "The <b>data</b> science team is the core of 6sense <b>engineering</b>. Our <b>data</b> scientists are not optimizing software; We’re growing quickly and want to expand our <b>data</b>...", "city": "San Francisco", "state": "CA", "date": "Mon, 18 Jul 2016 05:47:44 GMT", "url": "http://www.indeed.com/viewjob?jk=231996de16f8c0a0&qd=w-2ovI1tcnsfYPwo0sLkH9zbhbnue5uotyVhKGbVyKMWZG2Qcqk4kcnEktztFnFoBz9U92JvRWUwZzl66iD44zxoxgsbBlaVUvGtwJ4oKEZSI4CmH9UzR8aTbSGPezp5qbUCPeLOQ7Dzs0cmDHOdgA&indpubnum=9693529091171604&atk=1astii41qburm805", "country": "US", "formattedLocation": "San Francisco, CA", "jobtitle": "Data Scientist", "company": "6sense", "formattedLocationFull": "San Francisco, CA"}""" :: Nil)
         //"""{"snippet": "The data science team is the core of 6sense engineering. Our data scientists are not optimizing software; We’re growing quickly and want to expand our data...", "city": "San Francisco", "state": "CA", "date": "Mon, 18 Jul 2016 05:47:44 GMT", "url": "http://www.indeed.com/viewjob?jk=231996de16f8c0a0&qd=w-2ovI1tcnsfYPwo0sLkH9zbhbnue5uotyVhKGbVyKMWZG2Qcqk4kcnEktztFnFoBz9U92JvRWUwZzl66iD44zxoxgsbBlaVUvGtwJ4oKEZSI4CmH9UzR8aTbSGPezp5qbUCPeLOQ7Dzs0cmDHOdgA&indpubnum=9693529091171604&atk=1astii41qburm805", "country": "US", "formattedLocation": "San Francisco, CA", "jobtitle": "Data Scientist", "company": "6sense", "formattedLocationFull": "San Francisco, CA"}""" :: Nil)
         //"""{"snippet": "", "city": "San Francisco", "state": "CA", "date": "Mon, 18 Jul 2016 05:47:44 GMT", "url": "http://www.indeed.com/viewjob?jk=231996de16f8c0a0&qd=w-2ovI1tcnsfYPwo0sLkH9zbhbnue5uotyVhKGbVyKMWZG2Qcqk4kcnEktztFnFoBz9U92JvRWUwZzl66iD44zxoxgsbBlaVUvGtwJ4oKEZSI4CmH9UzR8aTbSGPezp5qbUCPeLOQ7Dzs0cmDHOdgA&indpubnum=9693529091171604&atk=1astii41qburm805", "country": "US", "formattedLocation": "San Francisco, CA", "jobtitle": "Data Scientist", "company": "6sense", "formattedLocationFull": "San Francisco, CA"}""" :: Nil)

      val diceStaticFile = sc.parallelize(
         """{"date": "2016-09-09", "jobTitle": "Data and Application Engineering Manager", "company": "EPE Innovations", "location": "Dallas, TX", "detailUrl": "http://www.dice.com/job/result/90955654/637266?src=19"}""" :: 
         """{"date": "2016-09-09", "jobTitle": "Data and Application Engineering Manager", "company": "EPE Innovations", "location": "Dallas, TX", "detailUrl": "http://www.dice.com/job/result/90955654/637266?src=19"}""" ::
        Nil)

      // read the files
   /**
      val indeedSchema = (new StructType).add("snippet", StringType).add("city", StringType).add("state", StringType).add("date", StringType).add("url", StringType).add("country", StringType).add("formattedLocation", StringType).add("jobtitle", StringType).add("company", StringType).add("formattedLocationFull", StringType)
      val diceSchema = (new StructType).add("date", StringType).add("jobTitle", StringType).add("company", StringType).add("location", StringType).add("detailUrl", StringType)
      val dfIndeed = csqlContext.read.schema(indeedSchema).json(indeedStaticFile)
      val dfDice = csqlContext.read.schema(diceSchema).json(diceStaticFile)
   */

      val indeedDF = csqlContext.read.json(indeedStaticFile)
      val diceDF = csqlContext.read.json(diceStaticFile)

      //val diceS3DF = csqlContext.read.json(diceFileS3)

      indeedDF.show
      diceDF.show
      indeedDF.printSchema()
      diceDF.printSchema()

      indeedDF.registerTempTable("indeedTBL")
      diceDF.registerTempTable("diceTBL")

      val newIndeedDF = csqlContext.sql("SELECT jobtitle, company, url, formattedLocation as location, date, snippet FROM indeedTBL")
      val newDiceDF = csqlContext.sql("SELECT jobTitle as jobtitle, company, detailUrl as url, location, date , detailUrl as snippet FROM diceTBL")
      newIndeedDF.registerTempTable("newIndeedTBL")
      newDiceDF.registerTempTable("newDiceTBL")

      // Join both DF contents
      //val combinedDF = csqlContext.sql("SELECT jobtitle, company, url, location, date, snippet INTO newDiceTBL FROM newIndeedTBL")
      val combinedDF = csqlContext.sql("SELECT jobtitle, company, url, location, date, snippet FROM newDiceTBL UNION ALL SELECT jobtitle, company, url, location, date, snippet FROM newIndeedTBL")
      combinedDF.registerTempTable("combinedTBL")

      // Dedup rows
      val combinedDedupDF = csqlContext.sql("SELECT jobtitle, company, first(url), location, first(date), first(snippet) FROM combinedTBL GROUP BY jobtitle, company, location")   
      combinedDedupDF.show

      combinedDedupDF.rdd.saveToEs("dashboard/jobs")
      //sc.makeRDD(combinedDedupDF).saveToEs("dashboard/jobs")
      val finalRDD = combinedDedupDF.rdd
      EsSpark.saveToEs(finalRDD, "dashboard/jobs")

      combinedDedupDF.show

      // Send Data to Elasticsearch	
      //combinedDedupDF.foreachRDD { rdd => {
	// rdd.saveToEs("dashboard/jobs")
	// }
      //}

      //EsSpark.saveJsonToEs(combinedDedupDF.rdd, "jobs/myApp")
      
      //val cRDD: org.apache.spark.rdd.RDD[org.apache.spark.sql.Row] = combinedDedupDF.rdd
      //cRDD.saveToEs("job")
      //val es_df = csqlContext.read.format("org.elasticsearch.spark.sql").load("jobs")
      //es_df.show
     
   }
}   

File: batch_pyspark/jobs_batch.py
import json
import pyspark
import os
from pyspark import SparkContext, SparkConf, StorageLevel
from pyspark.sql import SQLContext
from pyspark.sql.types import *
from elasticsearch import Elasticsearch
from datetime import datetime

#----------------------- Define constants ----------------------------#

#Spark
#SPARK_MASTER = os.environ['SPARK_MASTER']
SPARK_MASTER = 'ec2-52-89-46-245.us-west-2.compute.amazonaws.com'

# Elasticsearch
#ES_NODES = os.environ['ES_NODES']
#ES_NODES = 'ec2-52-26-9-10.us-west-2.compute.amazonaws.com:9200'

ES_NODES = 'ec2-52-26-9-10.us-west-2.compute.amazonaws.com'
#ES_NODES = ['ec2-52-26-9-10.us-west-2.compute.amazonaws.com:9200','ec2-54-68-213-131.us-west-2.compute.amazonaws.com:9200','ec2-52-43-52-129.us-west-2.compute.amazonaws.com:9200'])
#ES_NODES = 'ec2-52-26-9-10.us-west-2.compute.amazonaws.com,ec2-54-68-213-131.us-west-2.compute.amazonaws.com,ec2-52-43-52-129.us-west-2.compute.amazonaws.com'


ES_INDEX = 'dashboard'
#ES_TYPE = 'jobposting'
ES_TYPE = 'jobpost'
#ES_RESOURCE = 'dashboard/jobposting'
ES_RESOURCE = 'dashboard/jobpost'
es = Elasticsearch([{'host': ES_NODES}])

#Amazon S3
S3_BUCKET = 'giovanna-insight'

# Shema structure


feedStruct  = [StructField("jobtitle", StringType(), True),
        StructField("company", StringType(), True),
        StructField("url", StringType(), True),
        StructField("location", StringType(), True),
        StructField("snippet", StringType(), True),
        StructField("strday", IntegerType(), True),
        StructField("strmonth", IntegerType(), True),
        StructField("stryear", IntegerType(), True),
        StructField("creationdate", StringType(), True),
        StructField("origin", StringType(), True)]

'''
indeedStruct  = [StructField("city", StringType(), True),
        StructField("company", StringType(), True),
        StructField("country", StringType(), True),
        StructField("date", StringType(), True),
        StructField("expired", BooleanType(), True),
        StructField("formattedLocation", StringType(), True),
        StructField("formattedLocationFull", StringType(), True),
        StructField("formattedRelativeTime", StringType(), True),
        StructField("indeedApply", BooleanType(), True),
        StructField("jobkey", StringType(), True),
        StructField("jobtitle", StringType(), True),
        StructField("onmousedown", StringType(), True),
        StructField("snippet", StringType(), True),
        StructField("source", StringType(), True),
        StructField("sponsored", BooleanType(), True),
        StructField("state", StringType(), True),
        StructField("url", StringType(), True)]


diceStruct  = [StructField("company", StringType(), True),
        StructField("date", StringType(), True),
        StructField("detailUrl", StringType(), True),
        StructField("jobTitle", StringType(), True),
        StructField("location", StringType(), True)]
'''


#result = es.search(index="dashboard", body={'query': {'match': {'jobtitle': 'data_engineering'}}})
#print json.dumps(result, indent=2)


#def notFoundInJobsDB(row):
def notFoundInJobsDB(jobtitle, company, location):

    criteria_list = []
    criteria_list.append({ "match": { "jobtitle": jobtitle }})
    criteria_list.append({ "match": { "company": company }})
    criteria_list.append({ "match": { "location": location }})

    body = {"query": { "bool": {  "must": criteria_list } } }
    res = es.search(index = ES_INDEX, doc_type = ES_TYPE, size = 500, body = body)
    if res.length:
        return false
    else:
        return true    


def create_es_index():
   es_settings = {'number_of_shards':3, 'number_of_replicas': 2, 'refresh_interval': '1s', 'index.translog.flush_threshold_size': '1gb'}
   es_mapping = {"jobposting": {"properties": {"jobtitle": { "type": "string" }, "company": { "type": "string" }, "url": { "type": "string" }, "location": { "type": "string" }, "snippet": { "type": "string" }, "strday": {"type":"string"}, "strmonth": {"type":"string"}, "stryear": {"type":"string"}, "creationdate": { "type ": "string"}, "origin": { "type": "string" }}}} 
   response = es.indices.create(index=ES_INDEX, body={'settings': es_settings, 'mappings': es_mapping})



if __name__ == '__main__':

   print "****************** Result = TEST  **********************"   
   #sc = SparkContext("local", "Jobs batch App")
   sc = SparkContext('spark://' + SPARK_MASTER + ':7077', 'jobs_batch')
   sqlContext = SQLContext(sc)

   if not es.indices.exists(ES_INDEX):
      create_es_index()

   # folder on HDFS and Amazon S3 to pull the data from
   #val diceFile = "hdfs://ec2-52-89-46-245.us-west-2.compute.amazonaws.com:9000/camus/exec/history/2016-09-16*"
   indeedFileS3 = "s3a://giovanna-insight/raw_logs/secor_backup/indeedFeed/*"
   diceFileS3 = "s3a://giovanna-insight/raw_logs/secor_backup/diceFeed/*"
       
   # construct RDD[Sting]
   indeedDF = sqlContext.read.json(indeedFileS3)
   diceDF = sqlContext.read.json(diceFileS3)

   indeedDF.printSchema()
   diceDF.printSchema()
   print indeedDF.rdd.take(1)
   print diceDF.rdd.take(1)


   # Transform data to extract only the elements that are needed
   feedSchema = StructType(feedStruct)

   dateFormat = "%a, %d %b %Y %H:%M:%S %Z"
 
   indeedRDD = indeedDF.map(lambda row: pyspark.sql.Row(jobtitle=row.jobtitle, \
                                                          company=row.company, \
                                                          url=row.url, \
                                                          location=row.formattedLocation, \
                                                          snippet=row.snippet, \
                                                          #day=datetime.utcfromtimestamp(float(row.date)).day, \
                                                          #month=datetime.utcfromtimestamp(float(row.date)).month, \
                                                          #year=datetime.utcfromtimestamp(float(row.created_utc)).year, \
                                                          strday=str(datetime.strptime(row.date, dateFormat).day), \
                                                          strmonth=str(datetime.strptime(row.date, dateFormat).month), \
                                                          stryear=str(datetime.strptime(row.date, dateFormat).year), \
                                                          creationdate=str(datetime.strptime(row.date, dateFormat).day).zfill(2) + '-' + str(datetime.strptime(row.date, dateFormat).month).zfill(2) + '-' +  str(datetime.strptime(row.date, dateFormat).year), \
                                                          origin='Indeed'))

   print indeedRDD.take(1)
   #transformedIndeedDF = sqlContext.createDataFrame(indeedRDD, feedSchema).persist(StorageLevel.DISK_ONLY)
   transformedIndeedDF = sqlContext.createDataFrame(indeedRDD)

   diceRDD = diceDF.map(lambda row: pyspark.sql.Row(jobtitle=row.jobTitle, \
                                                          company=row.company, \
                                                          url=row.detailUrl, \
                                                          location=row.location, \
                                                          snippet='', \
                                                          strday=row.date[8:10], \
                                                          strmonth=row.date[5:7], \
                                                          stryear=row.date[0:4], \
                                                          creationdate=row.date[8:10] + '-' + row.date[5:7] + '-' + row.date[0:4], \
                                                          origin='Dice'))

   print diceRDD.take(1)

   #transformedDiceDF = sqlContext.createDataFrame(diceRDD, feedSchema).persist(StorageLevel.DISK_ONLY)
   transformedDiceDF = sqlContext.createDataFrame(diceRDD)

   print "--------------------------------transformedIndeedDF schema -------------------------------"
   transformedIndeedDF.printSchema()
   print "--------------------------------transformedDiceDF schema -------------------------------"
   transformedDiceDF.printSchema()
   print "--------------------------------indeed and Dice data -------------------------------"
   print transformedIndeedDF.rdd.take(1)
   print transformedDiceDF.rdd.take(1)

   

   transformedIndeedDF.registerTempTable("newIndeedTBL")
   transformedDiceDF.registerTempTable("newDiceTBL")

   # Join both DF contents
   combinedDF = sqlContext.sql("SELECT jobtitle, company, url, location, snippet, strday, strmonth, stryear, creationdate, origin FROM newDiceTBL UNION ALL SELECT jobtitle, company, url, location, snippet, strday, strmonth, stryear, creationdate, origin FROM newIndeedTBL")
   combinedDF.registerTempTable("combinedTBL")


   # Dedupe rows
   combinedDedupeDF = sqlContext.sql("SELECT jobtitle, company, location, first(url) as url, first(snippet) as snippet, first(strday) as strday, first(strmonth) as strmonth, first(stryear) as stryear, first(creationdate) as creationdate, first(origin) as origin FROM combinedTBL GROUP BY jobtitle, company, location")   

   combinedDedupeDF.printSchema() 
   print combinedDedupeDF.rdd.take(10)


   es_conf = {'es.nodes': ES_NODES, 'es.resource': ES_RESOURCE, 'es.port' : '9200',  'es.batch.write.retry.count': '-1', 'es.batch.size.bytes': '0.05mb'}


   finalRDD = combinedDedupeDF.rdd.map(lambda row: ('key', row.asDict()))
   #combinedDedupeRDD = combinedDedupeDF.filter(notFoundInJobsDB(combinedDedupeDF.jobtitle, combinedDedupeDF.company, combinedDedupeDF.location))
   #combinedDedupeDF = combinedDedupeDF.filter(notFoundInJobsDB(combinedDedupeDF.jobtitle, combinedDedupeDF.company, combinedDedupeDF.location))
   #finalRDD = combinedDedupeDF.rdd.map(lambda row: ('key', row.asDict()))

   print finalRDD.take(10)


   #finalRDD.saveAsNewAPIHadoopFile(path='-', \
   #                                         outputFormatClass='org.elasticsearch.hadoop.mr.EsOutputFormat', \
   #                                         keyClass='org.apache.hadoop.io.NullWritable', \
   #                                         valueClass='org.elasticsearch.hadoop.mr.LinkedMapWritable', \
   #                                         conf=es_conf)
 

   # Store distinct companies
   combinedDedupeDF.registerTempTable("dedupedTBL")
   companiesDF = sqlContext.sql("SELECT distinct company FROM dedupedTBL")   
   companiesRDD = companiesDF.rdd
   #filteredCompaniesRDD = companiesRDD.filter(lambda x: notFoundInDB(x))
   #filteredCompaniesRDD = filteredCompaniesRDD.rdd.map(lambda row: ('key', row.asDict()))

   #es_company_conf = {'es.nodes': ES_NODES, 'es.resource': 'dashboard/company', 'es.port' : '9200',  'es.batch.write.retry.count': '-1', 'es.batch.size.bytes': '0.05mb'}

   #filteredCompaniesRDD.saveAsNewAPIHadoopFile(path='-', \
   #                                         outputFormatClass='org.elasticsearch.hadoop.mr.EsOutputFormat', \
   #                                         keyClass='org.apache.hadoop.io.NullWritable', \
   #                                         valueClass='org.elasticsearch.hadoop.mr.LinkedMapWritable', \
   #                                         conf=es_company_conf)


